exporter/loadbalancingexporter/resolver_srv_test.go

```diff
@@ -0,0 +1,463 @@
+				{Target: "pod-1.service-1.ns.svc.cluster.local", Port: 4317, Priority: 0, Weight: 0},
+				{Target: "pod-2.service-1.ns.svc.cluster.local", Port: 4317, Priority: 0, Weight: 0},
+			}, nil
+		},
+	}
+
+	// test
+	var resolved []string
+	res.onChange(func(endpoints []string) {
+		resolved = endpoints
+	})
+	require.NoError(t, res.start(context.Background()))
+	defer func() {
+		require.NoError(t, res.shutdown(context.Background()))
+	}()
+
+	// verify
+	assert.Len(t, resolved, 3)
+	for i, value := range []string{
+		"pod-0.service-1.ns.svc.cluster.local",
+		"pod-1.service-1.ns.svc.cluster.local",
+		"pod-2.service-1.ns.svc.cluster.local",
+	} {
+		assert.Equal(t, value, resolved[i])
+	}
+}
+
+func TestInitialSRVResolutionWithPort(t *testing.T) {
+	// prepare
+	res, err := newSRVResolver(zap.NewNop(), "_port._proto.service-1.ns.svc.cluster.local", "12000", 5*time.Second, 1*time.Second)
+	require.NoError(t, err)
+
+	res.resolver = &mockSRVResolver{
+		onLookupIPAddr: func(context.Context, string) ([]net.IPAddr, error) {
+			return []net.IPAddr{
+				{IP: net.IPv4(12, 12, 12, 12)},
+			}, nil
+		},
+		onLookupSRV: func(context.Context, string, string, string) (string, []*net.SRV, error) {
+			return "notapplicable", []*net.SRV{
+				{Target: "pod-0.service-1.ns.svc.cluster.local", Port: 4317, Priority: 0, Weight: 0},
+				{Target: "pod-1.service-1.ns.svc.cluster.local", Port: 4317, Priority: 0, Weight: 0},
+				{Target: "pod-2.service-1.ns.svc.cluster.local", Port: 4317, Priority: 0, Weight: 0},
+			}, nil
+		},
+	}
+
+	// test
+	var resolved []string
+	res.onChange(func(endpoints []string) {
+		resolved = endpoints
+	})
+	require.NoError(t, res.start(context.Background()))
+	defer func() {
+		require.NoError(t, res.shutdown(context.Background()))
+	}()
+
+	// verify
+	assert.Len(t, resolved, 3)
+	for i, value := range []string{
+		"pod-0.service-1.ns.svc.cluster.local:12000",
+		"pod-1.service-1.ns.svc.cluster.local:12000",
+		"pod-2.service-1.ns.svc.cluster.local:12000",
+	} {
+		assert.Equal(t, value, resolved[i])
+	}
+}
+
+func TestErrNoSRVHostname(t *testing.T) {
+	// test
+	res, err := newSRVResolver(zap.NewNop(), "", "", 5*time.Second, 1*time.Second)
+
+	// verify
+	assert.Nil(t, res)
+	assert.Equal(t, errNoSRV, err)
+}
+
+// This tests the case where a failure to resolve does not halt the resolver.
+func TestFailureToResolve(t *testing.T) {
+	// prepare
+	res, err := newSRVResolver(zap.NewNop(), "_port._proto.service-1.ns.svc.cluster.local", "", 5*time.Second, 1*time.Second)
+	require.NoError(t, err)
+
+	expectedErr := errors.New("underlying resolution error")
+	res.resolver = &mockSRVResolver{
+		onLookupSRV: func(context.Context, string, string, string) (string, []*net.SRV, error) {
+			return "", nil, expectedErr
+		},
+	}
+
+	// test
+	require.NoError(t, res.start(context.Background()))
+
+	// verify
+	assert.NoError(t, err)
+}
+
+func TestMultipleIPsOnATarget(t *testing.T) {
```

Comment:
I don't remember 100% of the SRV record spec, but aren't hostnames allowed to be regular CNAME hostnames, backed by multiple A records?

Reply:
It does work in most cases but it's actually explicitly stated it should not be a CNAME in the [rfc](https://www.ietf.org/rfc/rfc2782.txt). It both defines "address records" as `A` and `AAAA` only and under `target` it states it must not be an alias and references RFC 1034 for alias which is the RFC talking about `CNAME`s.